File: app/src/main/java/net/ktnx/mobileledger/model/LedgerAccount.java
/*
 * Copyright © 2024 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import net.ktnx.mobileledger.db.Account;
import net.ktnx.mobileledger.db.AccountValue;
import net.ktnx.mobileledger.db.AccountWithAmounts;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class LedgerAccount {
    private static final char ACCOUNT_DELIMITER = ':';
    static Pattern reHigherAccount = Pattern.compile("^[^:]+:");
    private final LedgerAccount parent;
    private long dbId;
    private long profileId;
    private String name;
    private String shortName;
    private int level;
    private boolean expanded;
    private List<LedgerAmount> amounts;
    private boolean hasSubAccounts;
    private boolean amountsExpanded;

    public LedgerAccount(String name, @Nullable LedgerAccount parent) {
        this.parent = parent;
        if (parent != null && !name.startsWith(parent.getName() + ":"))
            throw new IllegalStateException(
                    String.format("Account name '%s' doesn't match parent account '%s'", name,
                            parent.getName()));
        this.setName(name);
    }
    @Nullable
    public static String extractParentName(@NonNull String accName) {
        int colonPos = accName.lastIndexOf(ACCOUNT_DELIMITER);
        if (colonPos < 0)
            return null;    // no parent account -- this is a top-level account
        else
            return accName.substring(0, colonPos);
    }
    public static boolean isParentOf(@NonNull String possibleParent, @NonNull String accountName) {
        return accountName.startsWith(possibleParent + ':');
    }
    @NonNull
    static public LedgerAccount fromDBO(AccountWithAmounts in, LedgerAccount parent) {
        LedgerAccount res = new LedgerAccount(in.account.getName(), parent);
        res.dbId = in.account.getId();
        res.profileId = in.account.getProfileId();
        res.setName(in.account.getName());
        res.setExpanded(in.account.isExpanded());
        res.setAmountsExpanded(in.account.isAmountsExpanded());

        res.amounts = new ArrayList<>();
        for (AccountValue val : in.amounts) {
            res.amounts.add(new LedgerAmount(val.getValue(), val.getCurrency()));
        }

        return res;
    }
    public static int determineLevel(String accName) {
        int level = 0;
        int delimiterPosition = accName.indexOf(ACCOUNT_DELIMITER);
        while (delimiterPosition >= 0) {
            level++;
            delimiterPosition = accName.indexOf(ACCOUNT_DELIMITER, delimiterPosition + 1);
        }
        return level;
    }
    @Override
    public int hashCode() {
        return name.hashCode();
    }
    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj == null)
            return false;

        if (!(obj instanceof LedgerAccount))
            return false;

        LedgerAccount acc = (LedgerAccount) obj;
        if (!name.equals(acc.name))
            return false;

        if (!getAmountsString().equals(acc.getAmountsString()))
            return false;

        return expanded == acc.expanded && amountsExpanded == acc.amountsExpanded;
    }
    // an account is visible if:
    //  - it has an expanded visible parent or is a top account
    public boolean isVisible() {
        if (parent == null)
            return true;

        return (parent.isExpanded() && parent.isVisible());
    }
    public boolean isParentOf(LedgerAccount potentialChild) {
        return potentialChild.getName()
                             .startsWith(name + ":");
    }
    private void stripName() {
        String[] split = name.split(":");
        shortName = split[split.length - 1];
        level = split.length - 1;
    }
    public String getName() {
        return name;
    }
    public void setName(String name) {
        this.name = name;
        stripName();
    }
    public void addAmount(float amount, @NonNull String currency) {
        if (amounts == null)
            amounts = new ArrayList<>();
        amounts.add(new LedgerAmount(amount, currency));
    }
    public void addAmount(float amount) {
        this.addAmount(amount, "");
    }
    public int getAmountCount() { return (amounts != null) ? amounts.size() : 0; }
    public String getAmountsString() {
        if ((amounts == null) || amounts.isEmpty())
            return "";

        StringBuilder builder = new StringBuilder();
        for (LedgerAmount amount : amounts) {
            String amt = amount.toString();
            if (builder.length() > 0)
                builder.append('\n');
            builder.append(amt);
        }

        return builder.toString();
    }
    public String getAmountsString(int limit) {
        if ((amounts == null) || amounts.isEmpty())
            return "";

        int included = 0;
        StringBuilder builder = new StringBuilder();
        for (LedgerAmount amount : amounts) {
            String amt = amount.toString();
            if (builder.length() > 0)
                builder.append('\n');
            builder.append(amt);
            included++;
            if (included == limit)
                break;
        }

        return builder.toString();
    }
    public int getLevel() {
        return level;
    }
    @NonNull
    public String getShortName() {
        return shortName;
    }
    public String getParentName() {
        return (parent == null) ? null : parent.getName();
    }
    public boolean hasSubAccounts() {
        return hasSubAccounts;
    }
    public void setHasSubAccounts(boolean hasSubAccounts) {
        this.hasSubAccounts = hasSubAccounts;
    }
    public boolean isExpanded() {
        return expanded;
    }
    public void setExpanded(boolean expanded) {
        this.expanded = expanded;
    }
    public void toggleExpanded() {
        expanded = !expanded;
    }
    public void removeAmounts() {
        if (amounts != null)
            amounts.clear();
    }
    public boolean amountsExpanded() {return amountsExpanded;}
    public void setAmountsExpanded(boolean flag) {amountsExpanded = flag;}
    public void toggleAmountsExpanded() {amountsExpanded = !amountsExpanded;}
    public void propagateAmountsTo(LedgerAccount acc) {
        for (LedgerAmount a : amounts)
            a.propagateToAccount(acc);
    }
    public boolean allAmountsAreZero() {
        for (LedgerAmount a : amounts) {
            if (a.getAmount() != 0)
                return false;
        }

        return true;
    }
    public List<LedgerAmount> getAmounts() {
        return amounts;
    }
    @NonNull
    public Account toDBO() {
        Account dbo = new Account();
        dbo.setName(name);
        dbo.setNameUpper(name.toUpperCase());
        dbo.setParentName(extractParentName(name));
        dbo.setLevel(level);
        dbo.setId(dbId);
        dbo.setProfileId(profileId);
        dbo.setExpanded(expanded);
        dbo.setAmountsExpanded(amountsExpanded);

        return dbo;
    }
    @NonNull
    public AccountWithAmounts toDBOWithAmounts() {
        AccountWithAmounts dbo = new AccountWithAmounts();
        dbo.account = toDBO();

        dbo.amounts = new ArrayList<>();
        List<LedgerAmount> amounts = getAmounts();
        if (amounts != null) {
            for (LedgerAmount amt : amounts) {
                AccountValue val = new AccountValue();
                val.setCurrency(amt.getCurrency());
                val.setValue(amt.getAmount());
                dbo.amounts.add(val);
            }
        }

        return dbo;
    }
    public long getId() {
        return dbId;
    }
}

File: app/src/main/kotlin/net/ktnx/mobileledger/model/LedgerAccount.kt
/*
 * Copyright © 2024 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.model

import net.ktnx.mobileledger.db.Account
import net.ktnx.mobileledger.db.AccountValue
import net.ktnx.mobileledger.db.AccountWithAmounts
import java.util.regex.Pattern

class LedgerAccount(
    name: String,
    private val parent: LedgerAccount?
) {
    private var dbId: Long = 0
    private var profileId: Long = 0
    var name: String = name
        set(value) {
            field = value
            stripName()
        }
    var shortName: String = ""
        private set
    var level: Int = 0
        private set
    var isExpanded: Boolean = false
    private var amounts: MutableList<LedgerAmount>? = null
    var hasSubAccounts: Boolean = false
    var amountsExpanded: Boolean = false

    // Java compatibility methods
    fun hasSubAccounts(): Boolean = hasSubAccounts
    fun amountsExpanded(): Boolean = amountsExpanded

    init {
        if (parent != null && !name.startsWith(parent.name + ":")) {
            throw IllegalStateException(
                String.format("Account name '%s' doesn't match parent account '%s'",
                    name, parent.name)
            )
        }
        this.name = name
    }

    val id: Long
        get() = dbId

    val parentName: String?
        get() = parent?.name

    // an account is visible if:
    //  - it has an expanded visible parent or is a top account
    val isVisible: Boolean
        get() = parent == null || (parent.isExpanded && parent.isVisible)

    val amountCount: Int
        get() = amounts?.size ?: 0

    private fun stripName() {
        val split = name.split(":")
        shortName = split.last()
        level = split.size - 1
    }

    fun isParentOf(potentialChild: LedgerAccount): Boolean {
        return potentialChild.name.startsWith("$name:")
    }

    @JvmOverloads
    fun addAmount(amount: Float, currency: String = "", amountStyle: AmountStyle? = null) {
        val amountList = amounts ?: ArrayList<LedgerAmount>().also { amounts = it }
        amountList.add(LedgerAmount(amount, currency, amountStyle))
    }

    fun getAmountsString(): String {
        val amts = amounts
        if (amts.isNullOrEmpty()) return ""

        return amts.joinToString("\n") { it.toString() }
    }

    fun getAmountsString(limit: Int): String {
        val amts = amounts
        if (amts.isNullOrEmpty()) return ""

        return amts.take(limit).joinToString("\n") { it.toString() }
    }

    fun toggleExpanded() {
        isExpanded = !isExpanded
    }

    fun removeAmounts() {
        amounts?.clear()
    }

    fun toggleAmountsExpanded() {
        amountsExpanded = !amountsExpanded
    }

    fun propagateAmountsTo(acc: LedgerAccount) {
        amounts?.forEach { it.propagateToAccount(acc) }
    }

    fun allAmountsAreZero(): Boolean {
        return amounts?.all { it.amount == 0f } ?: true
    }

    fun getAmounts(): List<LedgerAmount>? = amounts

    fun toDBO(): Account {
        val dbo = Account()
        dbo.name = name
        dbo.nameUpper = name.uppercase()
        dbo.parentName = extractParentName(name)
        dbo.level = level
        dbo.id = dbId
        dbo.profileId = profileId
        dbo.expanded = isExpanded
        dbo.amountsExpanded = amountsExpanded
        return dbo
    }

    fun toDBOWithAmounts(): AccountWithAmounts {
        val dbo = AccountWithAmounts()
        dbo.account = toDBO()

        val amountsList = ArrayList<AccountValue>()
        amounts?.forEach { amt ->
            val accountValue = AccountValue()
            accountValue.currency = amt.currency ?: ""
            accountValue.value = amt.amount
            // Save amount style if present
            amt.amountStyle?.let {
                accountValue.amountStyle = it.serialize()
            }
            amountsList.add(accountValue)
        }
        dbo.amounts = amountsList

        return dbo
    }

    override fun hashCode(): Int = name.hashCode()

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other !is LedgerAccount) return false

        if (name != other.name) return false
        if (getAmountsString() != other.getAmountsString()) return false

        return isExpanded == other.isExpanded && amountsExpanded == other.amountsExpanded
    }

    companion object {
        private const val ACCOUNT_DELIMITER = ':'
        private val reHigherAccount: Pattern = Pattern.compile("^[^:]+:")

        @JvmStatic
        fun extractParentName(accName: String): String? {
            val colonPos = accName.lastIndexOf(ACCOUNT_DELIMITER)
            return if (colonPos < 0) null else accName.substring(0, colonPos)
        }

        @JvmStatic
        fun isParentOf(possibleParent: String, accountName: String): Boolean {
            return accountName.startsWith("$possibleParent:")
        }

        @JvmStatic
        fun fromDBO(input: AccountWithAmounts, parent: LedgerAccount?): LedgerAccount {
            val res = LedgerAccount(input.account.name, parent)
            res.dbId = input.account.id
            res.profileId = input.account.profileId
            res.name = input.account.name
            res.isExpanded = input.account.expanded
            res.amountsExpanded = input.account.amountsExpanded

            val amountsList = ArrayList<LedgerAmount>()
            for (accountValue in input.amounts) {
                amountsList.add(LedgerAmount.fromDBO(accountValue))
            }
            res.amounts = amountsList

            return res
        }

        @JvmStatic
        fun determineLevel(accName: String): Int {
            var level = 0
            var delimiterPosition = accName.indexOf(ACCOUNT_DELIMITER)
            while (delimiterPosition >= 0) {
                level++
                delimiterPosition = accName.indexOf(ACCOUNT_DELIMITER, delimiterPosition + 1)
            }
            return level
        }
    }
}
